File: cpp/Construct-Binary-Tree-From-Ineorder-And-Postorder-Traversal/Construct-Binary-Tree-From-Ineorder-And-Postorder-Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        if (inorder.empty() || postorder.empty())
            return NULL;
        int n = inorder.size();
        int root = postorder[n-1];
        TreeNode* t = new TreeNode(root);
        if (n==1) return t;
        vector<int> left_in, left_post, right_in, right_post;
        int root_index = 0;
        for (int i=0;i<n;i++)
        {
            if (inorder[i] == root)
            {
                root_index=i;
                break;
            }
        }
        for (int i=0;i<root_index;i++)
        {
            left_in.push_back(inorder[i]);
            left_post.push_back(postorder[i]);
        }
        for (int i=root_index+1;i<n;i++)  // +1是跳过根节点
        {
            right_in.push_back(inorder[i]);
        }
        for (int i =root_index;i<n-1;i++)  // -1是跳过根节点
        {
            right_post.push_back(postorder[i]);
        }
        t->left = buildTree(left_in, left_post);
        t->right = buildTree(right_in, right_post);
        return t;
    }
};

File: python/Construct-Binary-Tree-From-Ineorder-And-Postorder-Traversal/Construct-Binary-Tree-From-Ineorder-And-Postorder-Traversal.py
### Solution 1:
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        if not inorder or not postorder:
            return None
        n = len(inorder)
        root = postorder[-1]
        t = TreeNode(root)
        for i in range(n):
            if inorder[i]==root:
                root_index = i
                break
        left_in = []
        left_post = []
        right_in = []
        right_post = []
        for i in range(root_index):
            left_in.append(inorder[i])
            left_post.append(postorder[i])
        for i in range(root_index+1,n):
            right_in.append(inorder[i])
        for i in range(root_index,n-1):
            right_post.append(postorder[i])
        t.left = self.buildTree(left_in, left_post)
        t.right = self.buildTree(right_in, right_post)
        return t
### Solution 2:
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        if not inorder or not postorder:
            return None
        root = TreeNode(postorder[-1])
        index = inorder.index(root.val)
        root.left = self.buildTree(inorder[:index],postorder[:index])
        root.right = self.buildTree(inorder[index+1:], postorder[index:-1])
        return root
